Best effort co-concretization (iterative algorithm)

alalazo opened the pull request:
**Use case description copied from #27964**

Currently, environments can either be concretized fully together or fully separately. This works well for users who create environments for interoperable software and can use concretization: together. It does not allow environments with conflicting software to be concretized for maximal interoperability.

The primary use-case for this is facilities providing system software. Facilities provide multiple MPI implementations, but all software built against a given MPI ought to be interoperable.

This PR adds a concretization option together_where_possible. When this option is used, Spack will concretize specs in the environment separately, but will optimize for minimal differences in overlapping packages.

**Implementation differences**

The algorithm used here is greedy, since specs are computed in multiple rounds where clingo concretizes together as many input specs as possible. The gist of the algorithm is:

https://github.com/alalazo/spack/blob/f6420eaf6eb5e10a3e4d0b0a1a6c20e4404f92c1/lib/spack/spack/solver/asp.py#L2092-L2135

To relax the requirements on input specs and allow for not solving some of them we add a new indirection:

https://github.com/alalazo/spack/blob/f6420eaf6eb5e10a3e4d0b0a1a6c20e4404f92c1/lib/spack/spack/solver/asp.py#L1787-L1802

and we give clingo the choice to solve for the input literal or not. All the input specs computed in previous rounds are then reused in later rounds to ensure the environment is as contained as possible.

**Comparison with #27964**
When trying out the encoding in #27964 for production environments we saw that there were huge requests:
- On memory during grounding, see https://github.com/spack/spack/pull/27964#issuecomment-1032474892
- On time, during solve, see https://github.com/spack/spack/pull/27964#discussion_r805136646

The solve-time can very likely be improved a lot by removing the symmetry of the new encoding with some artificial rules. The memory requests during the grounding phase instead will necessarily grow linearly with the number of PSIDs needed to solve the environment. That number is not known a-priori. It is capped by the number of specs in an environment (which is computationally unfeasible) but practically is usually much less. For instance, `e4s` has 119 specs and needs 3 process spaces to concretize.

This PR instead takes a greedy approach and perform N solves each with a single process space. It might give a different (and "suboptimal" i.e. with a number of subprocesses greater than strictly needed) result if compared with the encoding in #27964 but it is complete, i.e. if there is at least a solution the algorithm will not give a false negative. On the bright side, due to its greedy nature, it's much faster when solving for environments (`e4s` is solved in 3 rounds in ~1 min.).

becker33 committed:
add tests for best-effort coconcretization

becker33 committed:
test concretization preferences with coconcretization

alalazo committed:
Add a level of indirection to root specs

This commit introduce the "literal" atom, which comes with
a few different "arities". The unary "literal" contains an
integer that id the ID of a spec literal. Other "literals"
contain information on the requests made by literal ID. For
instance zlib@1.2.11 generates the following facts:

literal(0,"root","zlib").
literal(0,"node","zlib").
literal(0,"node_version_satisfies","zlib","1.2.11").

This should help with solving large environments "together
where possible" since later literals can be now solved
together in batches.

alalazo committed:
Add a mechanism to relax the number of literals being solved

becker33 committed:
Modify spack solve to display the new criteria

Since the new criteria is above all the build criteria,
we need to modify the way we display the output.

Originally done by Greg in #27964 and cherry-picked
to this branch by the co-author of the commit.

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Inject reusable specs into the solve

Instead of coupling the PyclingoDriver() object with
spack.config, inject the concrete specs that can be
reused.

A method level function takes care of reading from
the store and the buildcache.

alalazo committed:
Add a function to solve input_specs in multiple rounds

alalazo committed:
spack solve: show output of multi-rounds

alalazo committed:
Environment can concretize "together where possible"

alalazo commented:
@becker33 @tgamblin [e4s-pr-generate](https://gitlab.spack.io/spack/spack/-/jobs/1739708) took just 2 mins and a single core using `together_where_possible` compared to [3mins. and 16 cores](https://gitlab.spack.io/spack/spack/-/jobs/1738770) when we concretize separately.

alalazo committed:
Fix unit tests cherry-picked from #27964

alalazo committed:
Gitlab: use together_where_possible in build-systems pipeline

alalazo committed:
Update command completion

becker33 commented on lib/spack/spack/cmd/solve.py:
```diff
@@ -68,12 +69,60 @@ def setup_parser(subparser):
     subparser.add_argument(
         '--stats', action='store_true', default=False,
         help='print out statistics from clingo')
+    subparser.add_argument(
+        '--multiple-rounds', action='store_true', default=False,
+        help='allow solving input specs in multiple rounds')
```
I think this option should be named after its effects, not its implementation. I don't love `--allow-differences`, but that's the direction I think we should go. The fact it's done in multiple round is an implementation detail.

alalazo replied:
Maybe we can just call it `-T, --together-where-possible` to match the concretization mode in environments?

alalazo replied:
See 5a54a5e

becker33 commented on lib/spack/spack/cmd/solve.py:
```diff
@@ -102,58 +151,37 @@ def solve(parser, args):
+        if not setup_only:
+            _process_result(result, show, required_format, kwargs)
+    else:
+        for idx, result in enumerate(solver.solve_in_rounds(
+                specs, out=None, models=models, timers=args.timers, stats=args.stats
+        )):
+            tty.msg("ROUND {0}".format(idx))
+            tty.msg("")
```
I think we should only show the round information when `spack solve --show output`

alalazo replied:
The default when concretizing together is to show also the optimization weights and those change at each round. So here I'm showing the same information + the specs that will be moved to the next round.

alalazo replied:
This is not as good as I would like, but see 032bc0f If it seems acceptable we can defer improving the last bit of `spack solve` to a following, smaller PR.

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -19,10 +19,54 @@
 :- not 1 { node_os(Package, _) } 1, node(Package).
 :- not 1 { node_target(Package, _) } 1, node(Package).
 
+% each virtual mist have at most one provider in the dag
```
```suggestion
% each virtual must have at most one provider in the dag
```

alalazo replied:
Fixed in 03e27c2

becker33 commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1376,3 +1376,65 @@ def test_sticky_variant_in_package(self):
+        assert len(concrete_specs) == expected
+
+    @pytest.mark.parametrize('specs,expected_spec,occurances', [
+        # This document the greediness of the algorithm. Since this environment needs
+        # to be solved in 3 rounds, by selecting versions of dependencies in the
+        # opposite way as the root (newer dependencies for older roots) we have to use
+        # 3 versions of libelf instead of 2.
+        (['libdwarf@20130729^libelf@0.8.10',
+          'libdwarf@20130207^libelf@0.8.12',
+          'libdwarf@20111030'], 'libelf@0.8.12', 1),
```
I don't understand why this requires three versions of libelf -- `libdwarf@20111030` should be able to reuse either of the other libelf implementations. Is the issue that the first round of concretization is concretizing `libdwarf@20111030 ^ libefl@0.8.13` so we don't back out of that decision?

alalazo replied:
Yes, I'll investigate why that happens though.

alalazo replied:
There was a bug on what was reused and another when reconstructing reused specs. They should both be fixed in 53a102c

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -409,15 +435,32 @@ def specs(self):
+    @property
+    def unsolved_specs(self):
+        """List of abstract input specs that were not solved."""
+        if not self._unsolved_specs:
```
```suggestion
        if self._unsolved_specs is None:
```
That way we don't recompute every time if we already know the list is empty.

alalazo replied:
Fixed in 03e27c2

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -409,15 +435,32 @@ def specs(self):
+
+    @property
+    def specs_by_input(self):
+        if not self._concrete_specs_by_input:
```
I don't think the same applies here because the only way to have an empty list is if it's unsat, but we might as well just in case.
```suggestion
        if self._concrete_specs_by_input is None:
```

alalazo replied:
Fixed in 03e27c2

becker33 commented on lib/spack/spack/solver/asp.py:
```diff
@@ -409,15 +435,32 @@ def specs(self):
-        # The specs were already computed, return them
-        if self._concrete_specs:
-            return self._concrete_specs
+        if not self._concrete_specs:
```
Same here
```suggestion
        if self._concrete_specs is None:
```

alalazo replied:
Fixed in 03e27c2

alalazo committed:
Fix minor issues

alalazo committed:
Fix reusing specs from previous rounds

This change injects the reusable specs in the
SpecBuilder and avoids retrieving them from
either the DB or the buildcache (making the
underlying abstractions less coupled).

It also fixes a bug where we were "reusing"
only the root specs and not the other nodes of
the DAG.

alalazo committed:
spack solve: change option name to "-T, --together-where-possible"

alalazo committed:
Improve printing to stdout

alalazo committed:
Add a unit test that concretize an environment with "together where possible"

alalazo committed:
Add documentation for "together where possible"

becker33 commented on lib/spack/spack/cmd/solve.py:
```diff
@@ -102,58 +151,40 @@ def solve(parser, args):
-                sys.stdout.write(
-                    spec.tree(color=sys.stdout.isatty(), **kwargs))
+                print("% END ROUND {0}\n".format(idx))
+            _process_result(result, show, required_format, kwargs)
```
```suggestion
            if not setup_only:
                _process_result(result, show, required_format, kwargs)
```

alalazo replied:
Done in 7cebbff

becker33 commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1389,3 +1389,67 @@ def test_do_not_invent_new_concrete_versions_unless_necessary(self):
+        assert len(concrete_specs) == expected
+
+    @pytest.mark.parametrize('specs,expected_spec,occurances', [
+        # The algorithm is greedy, and it might decide to solve the "best"
+        # spec early in which case reuse is suboptimal. In this case the most
+        # recent version of libdwarf is selected and concretized to libelf@0.8.13
+        (['libdwarf@20111030^libelf@0.8.10',
+          'libdwarf@20130207^libelf@0.8.12',
+          'libdwarf@20130729'], 'libelf@0.8.12', 1),
```
Could we improve this by having both root literals (the literals we already use) and dependency literals (explicitly mentioned dependencies of roots) and minimizing the unsolved dependency literals as well? I think that would cause us to concretize the most specific specs first, which would increase our odds of a more minimal build.

alalazo replied:
Adding a comment, following a discussion with @becker33 yesterday. We can do something very similar to this by adding weights during the setup phase for the literals that are solved. Then we can minimize the sum of the weights for reuse (currently weights are uniform and equal to 1 for each literal spec).

We can experiment with different ways of computing the weights in a separate PR.

becker33 reviewed (changes requested):
One request, one idea

alalazo committed:
Handle setup-only case for best-effort co-concretization

alalazo commented:
@becker33 Pending unrelated errors in pipelines I think this is done.

alalazo commented:
@becker33 Is this PR waiting further review before being merged?

alalazo commented:
@becker33  It's a mystery to me how @eugeneswalker (or more likely Github on his behalf) could dismiss your review with a commit that doesn't exist anymore, but can you please :heavy_check_mark: again? 

I rebased the PR to double check that everything is working correctly with #28673 in `develop`.

eugeneswalker commented:
> @becker33 It's a mystery to me how @eugeneswalker (or more likely Github on his behalf) could dismiss your review with a commit that doesn't exist anymore, but can you please ✔️ again?

What da!?! I definitely didn't mean to do this, didn't even know about this PR. If anyone knows how this could happen...

alalazo commented:
@tgamblin Rebased and ready for review.

alalazo committed:
Fix unit-test checking two steps installation of specs with conflicting build-deps

alalazo committed:
Fix after rebase for error handling

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -7,6 +7,47 @@
+% Give clingo the choice to solve an input spec or not
+{ literal_solved(ID) } :- literal(ID).
+literal_not_solved(ID) :- not literal_solved(ID), literal(ID).
+
```
I think we can constrain this in asp to ensure at least one spec is concretized, that will get us better error messages than doing it in python.

```
:- C = #count{ ID : literal_solved(ID) }, C<1.
```

alalazo replied:
I agree with that since we now have #30669 Good catch, since this was not pointed out by a conflict during rebase!

alalazo replied:
Should be solved in 7c9ab7156fcf6c0ce905ee3f782760a0da0829e3

alalazo committed:
Enforce having at least a literal being solved

becker33 reviewed (commented):
I think this is ready, except that it will have conflicts with #30038 that will need to be resolved.

alalazo committed:
Resolve conflicts on configure option

becker33 commented on lib/spack/spack/cmd/solve.py:
```diff
@@ -70,7 +70,7 @@ def setup_parser(subparser):
         help='print out statistics from clingo')
     subparser.add_argument(
-        '-T', '--together-where-possible', action='store_true', default=False,
+        '-T', '--unify-when-possible', action='store_true', default=False,
```
This argument should be removed, we can just rely on the config option now (or make it a config-pushing option, like `--reuse`.

alalazo replied:
Fixed in 08bc55e

becker33 commented on lib/spack/spack/environment/environment.py:
```diff
@@ -773,9 +773,16 @@ def _read_manifest(self, f, raw_yaml=None):
 
         # Let `concretization` overrule `concretize:unify` config for now.
-        unify = spack.config.get('concretizer:unify')
+        unify = spack.config.get('concretizer:unify', 'separately')
+        default_concretization = unify
+        if unify is True:
+            default_concretization = 'together'
+        elif unify is False:
+            default_concretization = 'separately'
```
```suggestion
        try:
            translation = {'together': True, 'separately': False}
            self.concretization = translation[configuration['concretization']]
        except KeyError:
            self.concretization = spack.config.get('concretizer:unify', 'separately')

becker33 replied:
I think this is a much cleaner way of letting the old way take precedence.

alalazo replied:
Fixed in d7f254c

alalazo committed:
Adapt code to new concretization option internally

alalazo committed:
Remove option to spack solve
